<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="date" content="2019-08-14" />

<meta name="progressive" content="true" />
<meta name="allow-skip" content="false" />

<title>Analysis of Rhythms</title>


<!-- highlightjs -->
<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



</head>

<body>



<div class="pageContent band">
<div class="bandContent page">

<div class="topics">

<div id="section-circadian-data" class="section level2">
<h2>Circadian data</h2>
<hr />
<p><br></p>
<blockquote>
<p>Circadian data is typically a series of observations in time (<strong>timeseries</strong>)</p>
</blockquote>
<p><img src="images/parameters.png" width="400" /></p>
<p><span class="math inline">\(A\)</span> = (absolute) amplitude</p>
<p><span class="math inline">\(m\)</span> = magnitude</p>
<p><span class="math inline">\(\tau\)</span> = intrinsic period</p>
<p><span class="math inline">\(\phi\)</span> = phase</p>
<p><span class="math inline">\(a = \frac{A}{m}\)</span> = relative amplitude: amplitude as a fraction of the magnitude</p>
<p><span class="math inline">\(f = \frac{1 + a}{1-a} = \frac{m + A}{m - A}\)</span> = fold-amplitude = ratio of peak to trough values<br/></p>
<div id="section-ex.-1" class="section level3">
<h3>Ex. 1</h3>
<p><em>Calculate relative amplitude and fold-amplitude and then do the same after changing <span class="math inline">\(m\)</span> from 12 to 24</em></p>
<em>Check that you get the same values for from two ways of calculating <span class="math inline">\(f\)</span></em>
<div class="tutorial-exercise" data-label="rhy_params" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="8">
<pre class="text"><code>A &lt;- 6
m &lt;- 12</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":false,"error":false,"message":true,"exercise.df_print":"paged","exercise.checker":"NULL"}</script>
</div>
<div class="tutorial-exercise-support" data-label="rhy_params-hint" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>a &lt;- A/m
f &lt;- (1+a)/(1-a)
a
f</code></pre>
</div>
<p>The true scale (<em>left</em>) provides the absolute amplitude, while the normalized scale (<em>right</em>) provides the relative amplitude and fold-amplitude.</p>
<p><img src="tutorial_files/figure-html/plot_sine-1.png" width="624" /></p>
</div>
</div>
<div id="section-waveforms" class="section level2">
<h2>Waveforms</h2>
<hr />
<p><br></p>
<blockquote>
<p>Rhythmicity only requires a repeating pattern approx. every 24h.</p>
</blockquote>
<p>In reality, many shapes are possible</p>
<p><img src="images/weird_rhythms.png" width="500" /></p>
<blockquote>
<p>The sinusoidal shape is only for mathematical convenience.</p>
</blockquote>
</div>
<div id="section-analysis-goals" class="section level2">
<h2>Analysis goals</h2>
<hr />
<p><br> <strong>Why</strong> do we do want to perform this analysis?</p>
<ol style="list-style-type: decimal">
<li><p>decide if the timeseries is rhythmic or not</p></li>
<li><p>if rhythmic, compute the rhythm parameters (<span class="math inline">\(A, a, m, \tau\)</span>)<br/> <br/></p></li>
</ol>
<blockquote>
<p>The ability to do this is greatly affected by the <em>experimental design</em> (more on this later)</p>
</blockquote>
</div>
<div id="section-sampling-scheme" class="section level2">
<h2>Sampling scheme</h2>
<hr />
<p><br><br> <img src="images/sampling_schemes.png" width="600" /> <br><br><br></p>
<p>Many common techniques, such as qPCR and western blots, can yield either type of data! <br></p>
</div>
<div id="section-genomic-data" class="section level2">
<h2>Genomic data</h2>
<hr />
<p><br></p>
<p>Let us begin with an example of <em>preprocessed</em> microarray data from the mouse liver (<span class="citation">Hughes et al. (2009)</span>). The data were collected in constant dark over 48h, one sample every hour. Mice were sacrified to obtain a sample – so we have an <strong>independent sampling</strong> scheme.</p>
<div class="tutorial-exercise-support" data-label="hughes_intro-setup" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>data(&quot;cycMouseLiverRNA&quot;, package = &quot;MetaCycle&quot;)
hughes_data &lt;- gather(cycMouseLiverRNA, time ,expr, -geneName)
hughes_data$time &lt;- as.numeric(gsub(&quot;CT&quot;,&quot;&quot;,hughes_data$time))</code></pre>
</div>
<div id="section-ex.-2" class="section level3">
<h3>Ex. 2</h3>
<em>The data is in a <code>data.frame</code> called <code>cycMouseLiverRNA</code>. View the dataset by typing the name of the dataset and “Run”</em>
<div class="tutorial-exercise" data-label="hughes_intro" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="2">
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":3,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":false,"error":false,"message":true,"exercise.df_print":"paged","exercise.checker":"NULL"}</script>
</div>
</div>
<div id="section-ex.-3" class="section level3">
<h3>Ex. 3</h3>
<em>To be able to plot using <code>ggplot2</code> one needs the above data in a different ‘long’ format. The <code>cycMouseLiverRNA</code> in the long form is stored in <code>hughes_data</code>.</em> <em>Plot a few of your favorite clock genes from <code>hughes_data</code>(probe names are in the leftmost column of the data)</em>
<div class="tutorial-exercise" data-label="hughes_plot1" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="2">
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":3,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":false,"error":false,"message":true,"exercise.df_print":"paged","exercise.setup":"hughes_intro-setup","exercise.checker":"NULL"}</script>
</div>
<div class="tutorial-exercise-support" data-label="hughes_plot1-hint-1" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># View the data to find the columns containing the time and expression values.
hughes_data</code></pre>
</div>
<div class="tutorial-exercise-support" data-label="hughes_plot1-hint-2" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># Plot the timeseries using qplot from ggplot2. `filter` is used to select data for the desired gene.
qplot(time, expr, 
      data=filter(hughes_data, geneName == &quot;Per2_1417602_at&quot;), 
      xlab = &quot;time (h)&quot;, ylab = &quot;expression&quot;)</code></pre>
</div>
</div>
<div id="section-ex.-4" class="section level3">
<h3>Ex. 4</h3>
<blockquote>
<p>It is usual to plot gene expression data in the <strong>log<sub>2</sub></strong> scale.</p>
</blockquote>
<p><em>Can you plot the same clock gene in the log<sub>2</sub> scale?</em></p>
<div class="tutorial-exercise" data-label="hughes_plot2" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="2">
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":3,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":false,"error":false,"message":true,"exercise.df_print":"paged","exercise.setup":"hughes_intro-setup","exercise.checker":"NULL"}</script>
</div>
<div class="tutorial-exercise-support" data-label="hughes_plot2-hint" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># Use the `log2` function to transform the expression values.
qplot(time, log2(expr),
      data=filter(hughes_data, geneName == &quot;Per2_1417602_at&quot;), 
      xlab = &quot;time (h)&quot;, ylab = &quot;log2 expression&quot;)</code></pre>
</div>
</div>
</div>
<div id="section-harmonic-regression" class="section level2">
<h2>Harmonic Regression</h2>
<hr />
<blockquote>
<p>We are going to first discuss detection of rhythms only. Estimation of amplitude and phase will come later.</p>
</blockquote>
<p>This is the simplest approach to rhythm analysis. A cosine curve is fit to the timeseries (<span class="math inline">\(y\)</span>)</p>
<p><span class="math display">\[ y(t) = m + A\cos\left(\frac{2\pi t}{\tau} - \phi\right)\]</span></p>
<p>Do you recognise the different parameters from <a href="#section-circadian-data">earlier</a>?</p>
<p>The above equation is exactly the same as this one:</p>
<p><span class="math display">\[ y(t) = m + c_1 \cos\left(\frac{2\pi t}{\tau}\right) + c_2 \sin\left(\frac{2\pi t}{\tau}\right)\]</span></p>
<p>If you can confidently fix <span class="math inline">\(\tau\)</span>, for e.g., <span class="math inline">\(\tau = 24\)</span> for mice or <span class="math inline">\(\tau = 22.5\)</span> for <em>N. crassa</em>, then this becomes a least-squares fitting exercise. You might be more familiar with this as linear regression or fitting a straight line through a cloud of points.</p>
<p>Since a model (the sine curve) is fit to the data is termed a <strong>parametric</strong> approach in statistics.</p>
<table>
<colgroup>
<col width="44%" />
<col width="4%" />
<col width="52%" />
</colgroup>
<thead>
<tr class="header">
<th>Advantages</th>
<th></th>
<th>Disadvantages</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Regression is very very fast</td>
<td></td>
<td>Assumes rhythms of interest have sine waveform (recall this <a href="#section-circadian-data">topic</a>)</td>
</tr>
<tr class="even">
<td>Works for many different types of data (independent and longitudinal sampling)</td>
<td></td>
<td>The noise in your data must be Gaussian and independent between samples. Outliers really can mess things up.</td>
</tr>
<tr class="odd">
<td>Unevenly-spaced or missing samples are not a problem</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>Amplitude and phase estimates are directly available</td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td>Easily used within standard tools for differential gene expression analysis, such as <code>limma/edgeR</code> and <code>DESeq2</code></td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>Very flexible. So other trends in data can be included</td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<div id="section-ex.-5" class="section level3">
<h3>Ex. 5</h3>
<em>Let us look at the same liver dataset from before.</em> <em>Load the data in the workspace</em> <em>View the first few lines of the dataset using <code>head</code> command</em>
<div class="tutorial-exercise" data-label="hughes_data" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="8">
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":false,"error":false,"message":true,"exercise.df_print":"paged","exercise.checker":"NULL"}</script>
</div>
<div class="tutorial-exercise-support" data-label="hughes_data-solution" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code># To see the list of datasets in this package
# data(package=&quot;learnRhythmanalysis&quot;)

# Load the dataset
data(&quot;hughes_2009_liver&quot;, package = &quot;learnRhythmanalysis&quot;)

# View the first few rows of the dataset
head(hughes_2009_liver)</code></pre>
</div>
</div>
<div id="section-ex.-6" class="section level3">
<h3>Ex. 6</h3>
<div class="tutorial-exercise-support" data-label="hughes_harmonic-setup" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>library(HarmonicRegression)
data(&quot;hughes_2009_liver&quot;, package = &quot;learnRhythmanalysis&quot;)</code></pre>
</div>
<p><em>Use the <code>harmonic.regression()</code> function in the package <code>HarmonicRegression</code> for rhythm analysis. Run the analysis on the <span class="citation">Hughes et al. (2009)</span> data and store the results in the <code>results</code> variable</em></p>
<em>View the contents of the results. It includes amplitudes, phases and p-values</em>
<div class="tutorial-exercise" data-label="hughes_harmonic" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":false,"error":false,"message":true,"exercise.df_print":"paged","exercise.checker":"NULL"}</script>
</div>
<div class="tutorial-exercise-support" data-label="hughes_harmonic-hint-1" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>## Step 1. Load the data (already did in the previous first exercise)

## Step 2. Make the data into a matrix of numbers. Move gene names to rownames/columnnames as the case may be.

## Step 3. Transpose the data (genes in columns and times in rows) as `harmonic.regression` needs this format.

## Step 4. Create a vector of time points to match samples to times (this information is in the column names of the original data)

## Step 5. Run the analysis with the required parameters.

## Step 6. View the results</code></pre>
</div>
<div class="tutorial-exercise-support" data-label="hughes_harmonic-hint-2" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>## Step 1. Load the data (already did in the previous first exercise)

## Step 2. Make the data into a matrix of numbers. Move gene names to rownames/columnnames as the case may be.

hughes_data &lt;- as.matrix(hughes_2009_liver[-1]) ## drop the gene symbols in the first column)
rownames(hughes_data) &lt;- hughes_2009_liver$GeneName ## matrix in R can have rownames and columnnames</code></pre>
</div>
<div class="tutorial-exercise-support" data-label="hughes_harmonic-hint-3" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>## Step 3. Transpose the data (genes in columns and times in rows) as `harmonic.regression` needs this format.

hughes_data &lt;- t(hughes_data) ## t() is a transpose operator

## Step 4. Create a vector of time points to match samples to times (this information is in the column names of the original data)

times &lt;- seq(1,48) # create a vector of sample times</code></pre>
</div>
<div class="tutorial-exercise-support" data-label="hughes_harmonic-hint-4" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>## Step 5. Run the analysis with the required parameters.

results &lt;- harmonic.regression(inputts = hughes_data, 
                               inputtime = times, 
                               Tau = 24, 
                               normalize = TRUE)

## Step 6. View the results

ls.str(results) # the same information is available from the help for the function, i.e., type ?harmonic.regression</code></pre>
</div>
<p>The important information in this list are <code>pars</code> (contains amplitudes and phase), <code>pvals</code> (p-values) and <code>qvals</code> (q-values).</p>
</div>
</div>
<div id="section-p-values-q-values" class="section level2">
<h2>P-values &amp; Q-values</h2>
<hr />
<p><br></p>
<p>We need to correct for multiple-testing.</p>
<p>Multiple testing drives the number of false positives up and must be controlled.</p>
<p>The multiple tests are the large number of transcripts we are testing the genome-wide dataset.</p>
<p>Normally, we correct the p-values for multiple testing to obtain <strong>q-values</strong> or <strong>adjusted p-values</strong> (there are some statistical differences between the two but for your purposes they are the same.)</p>
<div id="section-ex.-7" class="section level3">
<h3>Ex. 7</h3>
<p><em>I have generated a matrix of random (normally distributed) values, i.e., there are no rhythms in this data.</em></p>
<p><em>First, perform harmonic regression</em></p>
<p><em>Then, find the number of timeseries with p-value &lt; 0.05</em></p>
<p><em>Finally, find the number of timeseries with adjusted p-value/q-value &lt; 0.05. What do you see?</em></p>
<div class="tutorial-exercise-support" data-label="noise_data-setup" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>library(HarmonicRegression)
data &lt;- matrix(rnorm(48*1000), ncol=48)</code></pre>
</div>
<div class="tutorial-exercise" data-label="noise_data" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>## View the random data if you like
# head(data)

## Show the dimension of the data (rows columns)
sprintf(&quot;The size of the dataset: %d&quot;, dim(data))

## Run harmonic regression on the dataset
results &lt;- harmonic.regression(t(data), 
                               inputtime = seq(1,48),
                               Tau = 24,
                               normalize = TRUE)</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":false,"error":false,"message":true,"exercise.df_print":"paged","exercise.checker":"NULL"}</script>
</div>
<div class="tutorial-exercise-support" data-label="noise_data-hint-1" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>## The number of &quot;genes&quot; with p-values less than 0.05
sum(results$pvals &lt; 0.05)</code></pre>
</div>
<div class="tutorial-exercise-support" data-label="noise_data-hint-2" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>## Correct for multiple-testing using Benjamini-Hochberg using `p.adjust`
## but harmonic regression already does this for you
sum(results$qvals &lt; 0.05)</code></pre>
</div>
<p><strong>Note</strong> The fraction of “rhythmic” timeseries found by the analysis is approximately <code>1000 * 0.05</code><br />
<br></p>
<hr />
<p><br></p>
<p><img src="tutorial_files/figure-html/noise_p_dist-1.png" width="576" /></p>
<p>On the other hand the p-value distribution on the real data we analyzed earlier, looks like <img src="tutorial_files/figure-html/hughes_p_dist-1.png" width="576" /></p>
<p>To learn about interpreting p-value distributions read <a href="http://varianceexplained.org/statistics/interpreting-pvalue-histogram/">this post</a>. <strong>Note</strong> these insights are only for the p-value distribution not for the q-values.</p>
<blockquote>
<p>In summary, if you <em>do not</em> see a p-value distribution that is flat except for a peak at the left end (near zero), then something is not okay with your <em>parametric </em> method.</p>
</blockquote>
</div>
</div>
<div id="section-non-parametric-approaches" class="section level2">
<h2>Non-parametric approaches</h2>
<hr />
<p><br></p>
<p>Two popular (and similar) approaches in this category are <code>JTKcycle</code> (<span class="citation">Hughes et al. (2010)</span>) and <code>rain</code> (<span class="citation">Thaben and Westermark (2014)</span>).</p>
<p>Both these methods relax the assumption about the strictly sinusoidal shape of the rhythms. More importantly, they only look for patterns of alternating increasing and decreasing values. So scale of the data (normal or log scale) does not matter.</p>
<div id="section-jtkcycle" class="section level3">
<h3>JTKcycle</h3>
<table>
<colgroup>
<col width="42%" />
<col width="3%" />
<col width="53%" />
</colgroup>
<thead>
<tr class="header">
<th>Advantages</th>
<th></th>
<th>Disadvantages</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Robust to outliers</td>
<td></td>
<td>Requires integer-intervals and even spacing between samples</td>
</tr>
<tr class="even">
<td>High computational efficiency</td>
<td></td>
<td>Dispersed output parameters (P-value, period and phase) for low resolution data</td>
</tr>
<tr class="odd">
<td>Improved power in analyzing data with duplicate samples</td>
<td></td>
<td>Cosine curve bias</td>
</tr>
<tr class="even">
<td>Works with missing data and replicates</td>
<td></td>
<td>False negative rate issue for low resolution timeseries</td>
</tr>
<tr class="odd">
<td>Does not assume specific noise distribution</td>
<td></td>
<td>Less accurate phase for low resolution timeseries</td>
</tr>
<tr class="even">
<td>Can test for a range of periods (<span class="math inline">\(\tau\)</span>)</td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</div>
<div id="section-ex.-8" class="section level3">
<h3>Ex. 8</h3>
<p><em>Apply JTKcycle on the <span class="citation">Hughes et al. (2009)</span> dataset</em> (<strong>Note</strong>: this is going to be slow, so be patient.)</p>
<em>View the results and the top 10 rhythmic genes</em>
<div class="tutorial-exercise-support" data-label="JTKcycle_run-setup" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>library(MetaCycle)
data(&quot;hughes_2009_liver&quot;, package = &quot;learnRhythmanalysis&quot;)</code></pre>
</div>
<div class="tutorial-exercise" data-label="JTKcycle_run" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>## Step 1. Load the data (already done).

## Step 2: Vector of sample times
times &lt;- seq(1,48)

## Step 3: Run meta2d (that has three methods JTKcycle, Lomb-Scargle and ARSER). 
## Here we will use only JTKcycle.
## The call is complex as it was originally only designed to work with files.
results &lt;- meta2d(inDF = hughes_2009_liver,
                  timepoints = times,
                  minper = 24,
                  maxper= 24,
                  cycMethod = &quot;JTK&quot;,
                  outputFile = FALSE,
                  infile = &quot;csv&quot;,
                  outdir =&quot;csv&quot;, 
                  filestyle = &quot;csv&quot;,
                  releaseNote = FALSE)</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":false,"error":false,"message":true,"exercise.df_print":"paged","exercise.checker":"NULL"}</script>
</div>
<div class="tutorial-exercise-support" data-label="JTKcycle_run-hint-1" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>## Step 4: View results -- the data is a list for each method chosen in the meta2d call. Here we are only looking at the JTK results
head(results$JTK)</code></pre>
</div>
<div class="tutorial-exercise-support" data-label="JTKcycle_run-hint-2" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>## Step 5: View the top 10 hits
head(results$JTK[order(results$JTK$ADJ.P),], n = 10)</code></pre>
</div>
</div>
<div id="section-rain" class="section level3">
<h3>RAIN</h3>
<table>
<colgroup>
<col width="42%" />
<col width="3%" />
<col width="53%" />
</colgroup>
<thead>
<tr class="header">
<th>Advantages</th>
<th></th>
<th>Disadvantages</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Unequal length of increasing and decreasing portions of the curve (removes any cosine curve bias)</td>
<td></td>
<td>Can be slow for long timeseries</td>
</tr>
<tr class="even">
<td>All advantages of JTKcycle</td>
<td></td>
<td>Underpowered for short timeseries as bootstrapping provides clustered p-values</td>
</tr>
<tr class="odd">
<td>P-values are estimated using bootstrapping</td>
<td></td>
<td>Clustered distribution of p-values can affect multiple-testing correction method</td>
</tr>
<tr class="even">
<td>Does not require integer-intervals between samples</td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</div>
<div id="section-ex.-9" class="section level3">
<h3>Ex. 9</h3>
<em>Now, we will test <code>rain</code> on RNA-seq data of the mouse liver. This dataset comes with <code>rain</code> package. It has six timepoints over 24h with two replicates at each timepoint.</em> <em>As always load the data and view it</em> <em>Run the RAIN analysis</em> <em>Remember to correct for multiple testing as RAIN does not do this for you</em>
<div class="tutorial-exercise-support" data-label="rain_run-setup" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>library(rain)
data(menetRNASeqMouseLiver, package=&quot;rain&quot;)</code></pre>
</div>
<div class="tutorial-exercise" data-label="rain_run" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>## Step 1: Load the data
data(menetRNASeqMouseLiver, package=&quot;rain&quot;)

head(menetRNASeqMouseLiver)</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":false,"error":false,"message":false,"exercise.df_print":"paged","exercise.checker":"NULL"}</script>
</div>
<div class="tutorial-exercise-support" data-label="rain_run-hint-1" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>## Step 2 &amp; 3: Transpose the data and run `rain` on the data
menet.ossc &lt;- rain(t( menetRNASeqMouseLiver ), 
                   deltat = 4, 
                   period = 24,
                   nr.series = 2, 
                   peak.border = c(0.3, 0.7))</code></pre>
</div>
<div class="tutorial-exercise-support" data-label="rain_run-hint-2" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>## Step 4: View the results
head(menet.ossc)</code></pre>
</div>
<div class="tutorial-exercise-support" data-label="rain_run-hint-3" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>## Step 5: Correct p-values for multiple testing using Benjamini-Hochberg and view the top genes
# the `transform` function lets you add new columns to the data.frame from the already present columns.
menet.ossc &lt;- transform(menet.ossc, 
                        adjusted.pVal = p.adjust(pVal, method = &quot;BH&quot;))

head(menet.ossc[order(menet.ossc$adjusted.pVal),], n=10)</code></pre>
</div>
<p><strong>Note:</strong> The identical p-values and adjusted p-values for many genes. This common in non-parametric tests – it will become less obvious with more data per timeseries.</p>
<blockquote>
<p>Please do <strong>not</strong> use phase or period estimates from <code>rain</code> or <code>JTKcycle</code>. We will come to parameter estimation next.</p>
</blockquote>
<blockquote>
<p>Both methods can only be applied on longitudinal timeseries from a single biological unit timeseries.</p>
</blockquote>
</div>
</div>
<div id="section-parameter-estimation" class="section level2">
<h2>Parameter estimation</h2>
<hr />
<p><br></p>
<p>The three tests discussed are used to detect rhythms (recall <a href="#section-analysis-goals">slide</a>). The standard approach is to follow <em>rhythm detection</em> by harmonic regression on the timeseries to obtain amplitudes and phases (<em>parameter estimation</em>).</p>
<p>Both detection and estimation are done together in the parametric harmonic regression. But the non-parametric methods are performed in two steps.</p>
<p>The <code>MetaCycle</code> package that contains the JTKcycle method performs these two steps and provides amplitude and phase estimates in the results. On the other hand, <code>rain</code> must be followed by <code>harmonic.regression</code> to get the amplitudes and phases.</p>
<p><strong>Note:</strong> <code>harmonic.regression</code> gives the relative and <em>not</em> the absolute amplitudes.</p>
<blockquote>
<p>It is common to follow a q-value-based selection of rhythmic timeseries by a relative amplitude-based filtering.</p>
</blockquote>
<p>The thinking behind this is as follows. The p-value provides a statistical measure of rhythmicity of a timeseries. It is not uncommon at all to find timeseries that have a very small amplitudes yet have very significant rhythms.</p>
<p>Such timeseries are statistically but <em>not biologically</em> significant. Therefore, we filter the rhythmic timeseries by ensuring that the relative-amplitude is greater than threshold (10%-15%).</p>
<blockquote>
<p>So statistics are important. But don’t get obsessed by the p-values/q-values. Biological relevance is equally important.</p>
</blockquote>
</div>
<div id="section-experimental-design" class="section level2">
<h2>Experimental design</h2>
<hr />
<p><br></p>
<p>As mentioned earlier, the ability to detect rhythms is extremely sensitive to the sampling of the timeseries.</p>
<blockquote>
<p>For very good rhythm detection, the current consensus is sampling every 2h over two circadian cycles (without replicates per sample). Having biological replicates at each timepoint would allow one to relax both of these requirements.</p>
</blockquote>
<blockquote>
<p>For estimating phases and amplitudes, higher sampling is preferred. Accuracy of phase is about half the sampling interval.</p>
</blockquote>
<p>For more information on this I refer you to the series of papers on this topic from the Hogenesch group (<span class="citation">Wu et al. (2014)</span>, <span class="citation">Li et al. (2015)</span>, <span class="citation">Hughes et al. (2017)</span>)</p>
</div>
<div id="section-workflow" class="section level2">
<h2>Workflow</h2>
<hr />
<p><br></p>
<p>Step 1. Organize your data in your computer or get public datasets from repositories (such as GEO using <code>GEOquery</code> package).</p>
<p>Step 2. Preprocess the data to a matrix of values. E.g., using <code>oligo</code> or <code>affy</code> packages for microsarray and alignment and counting using <code>Rsubread</code>.</p>
<p>Step 3. Normalize the data according to the technology for inter-sample variation</p>
<p>Step 4. Exploratory analysis (PCA, MDS plots) to check for batch effects, bad samples etc.</p>
<p>Step 5. Filtering out lowly-expressed genes based on mean/median expression.</p>
<p>Step 6. Rhythm analysis for detecting rhythms and estimating parameters.</p>
<p>Step 7. Select rhythmic genes according q-value and optionally the relative amplitude.</p>
<p>Step 8. Plot the results, perform GO analysis, ….</p>
</div>
<div id="section-smaller-datasets" class="section level2">
<h2>Smaller datasets</h2>
<hr />
<p><br></p>
<p>Often experiments in the lab produce smaller numbers of timeseries. All the discussed methods can also be used on such data.</p>
<p>However, one must bear in mind the disadvantages of the the particular method when you have few time samples/replicates. Also, since one performs fewer tests, one can opt for more conservative multiple-correction procedures, such as Boneferroni.</p>
<p>With very coarsely sampled data, the non-parametric methods are better as they make fewer assumptions.</p>
<p>The most statistically powerful method (when all else fails/one should start with this) is <strong>one-way ANOVA</strong> with time as the factor. If this is significant we known that measured quantity is time-dependent <strong>not</strong> that is rhythmic.</p>
</div>
<div id="section-golden-rules" class="section level2">
<h2>Golden Rules</h2>
<p><strong>1.</strong> Always look at the data. Seeing is believing. If you cannot see the rhythms it is probably not there.</p>
<p><strong>2.</strong> Include all effects in your analysis, e.g., different batches, different cultures. Check if they matter before ignoring them.</p>
<p><strong>3.</strong> Think about the analysis before you run experiments. If needed, get in touch with biostatistician/bioinformatician.</p>
<p><strong>4.</strong> P-value/Q-value &gt; threshold does not mean the timeseries is not rhythmic, only that the evidence is inconclusive. Absence of evidence is <strong>not</strong> evidence of absence.</p>
<p><strong>5.</strong> Never duplicate and concatenate data before running statistics (<span class="citation">Hughes et al. (2017)</span>)</p>
<p><strong>6.</strong> Control for multiple testing (<span class="citation">Hughes et al. (2017)</span>)</p>
<p><strong>7.</strong> Deposit your raw data into public repositories (<span class="citation">Hughes et al. (2017)</span>)</p>
<p><strong>8.</strong> Never directly alter the raw data. Always save any processed data separately.</p>
<p><strong>9.</strong> Never do two tests when you can get away with one – do differential rhythmicity analysis when you compare two rhythms (<span class="citation">Thaben and Westermark (2016)</span>)</p>
</div>
<div id="section-references-and-resources" class="section level2">
<h2>References and Resources</h2>
<p>Learn programming in R with free <a href="https://r4ds.had.co.nz">book</a> | <img src="images/cover_wickham.png" width="100" /></p>
<p><a href="https://cran.r-project.org/web/packages/MetaCycle/vignettes/implementation.html">MetaCycle</a> package for running three different rhythm analysis tools</p>
<p>Well-managed repositories for biological data analysis <a href="https://www.bioconductor.org/">Bioconductor</a>. They also have many workflows and tools to make your life easier.</p>
<p>Exploratory analysis and pre-processing (before rhythm analysis) workflows for microarrays (on <a href="https://www.bioconductor.org/packages/release/workflows/vignettes/arrays/inst/doc/arrays.html">bioconductor</a>) and RNA-seq (<span class="citation">Love et al. (2015)</span>)</p>
<p>Designing and conducting circadian RNA-seq study (<span class="citation">Li et al. (2015)</span>)</p>
<p><a href="https://biodare2.ed.ac.uk/">BioDare</a> – Fast period analysis, timeseries processing and aesthetic visualizations</p>
<div id="section-references" class="section level3">
<h3>References</h3>

<script type="application/shiny-prerendered" data-context="server-start">
library(learnr)
knitr::opts_chunk$set(echo = FALSE, warning = FALSE)
library(tidyr)
library(dplyr)
library(gridExtra)
library(ggplot2)
theme_set(theme_bw())
options(shiny.port=2123, tutorial.exercise.timelimit=180)
</script>
 
<script type="application/shiny-prerendered" data-context="server">
learnr:::register_http_handlers(session, metadata = NULL)
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-rhy_params-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-rhy_params-code-editor`)), session)
output$`tutorial-exercise-rhy_params-output` <- renderUI({
  `tutorial-exercise-rhy_params-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-hughes_intro-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-hughes_intro-code-editor`)), session)
output$`tutorial-exercise-hughes_intro-output` <- renderUI({
  `tutorial-exercise-hughes_intro-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-hughes_plot1-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-hughes_plot1-code-editor`)), session)
output$`tutorial-exercise-hughes_plot1-output` <- renderUI({
  `tutorial-exercise-hughes_plot1-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-hughes_plot2-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-hughes_plot2-code-editor`)), session)
output$`tutorial-exercise-hughes_plot2-output` <- renderUI({
  `tutorial-exercise-hughes_plot2-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-hughes_data-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-hughes_data-code-editor`)), session)
output$`tutorial-exercise-hughes_data-output` <- renderUI({
  `tutorial-exercise-hughes_data-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-hughes_harmonic-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-hughes_harmonic-code-editor`)), session)
output$`tutorial-exercise-hughes_harmonic-output` <- renderUI({
  `tutorial-exercise-hughes_harmonic-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-noise_data-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-noise_data-code-editor`)), session)
output$`tutorial-exercise-noise_data-output` <- renderUI({
  `tutorial-exercise-noise_data-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-JTKcycle_run-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-JTKcycle_run-code-editor`)), session)
output$`tutorial-exercise-JTKcycle_run-output` <- renderUI({
  `tutorial-exercise-JTKcycle_run-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-rain_run-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-rain_run-code-editor`)), session)
output$`tutorial-exercise-rain_run-output` <- renderUI({
  `tutorial-exercise-rain_run-result`()
})
</script>
 <!--html_preserve-->
<script type="application/shiny-prerendered" data-context="dependencies">
{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["jquery"]},{"type":"character","attributes":{},"value":["1.11.3"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/jquery"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquery.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.14"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["bootstrap"]},{"type":"character","attributes":{},"value":["3.3.5"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/bootstrap"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["viewport"]}},"value":[{"type":"character","attributes":{},"value":["width=device-width, initial-scale=1"]}]},{"type":"character","attributes":{},"value":["js/bootstrap.min.js","shim/html5shiv.min.js","shim/respond.min.js"]},{"type":"character","attributes":{},"value":["css/lumen.min.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.14"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["pagedtable"]},{"type":"character","attributes":{},"value":["1.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/pagedtable-1.1"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["js/pagedtable.js"]},{"type":"character","attributes":{},"value":["css/pagedtable.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.14"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["highlightjs"]},{"type":"character","attributes":{},"value":["9.12.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/highlightjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["highlight.js"]},{"type":"character","attributes":{},"value":["textmate.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.14"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial"]},{"type":"character","attributes":{},"value":["0.9.2.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial.js"]},{"type":"character","attributes":{},"value":["tutorial.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-autocompletion"]},{"type":"character","attributes":{},"value":["0.9.2.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-autocompletion.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-diagnostics"]},{"type":"character","attributes":{},"value":["0.9.2.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-diagnostics.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-format"]},{"type":"character","attributes":{},"value":["0.9.2.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmarkdown/templates/tutorial/resources"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-format.js"]},{"type":"character","attributes":{},"value":["tutorial-format.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["jquery"]},{"type":"character","attributes":{},"value":["1.11.3"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/jquery"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquery.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.14"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["font-awesome"]},{"type":"character","attributes":{},"value":["5.1.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/fontawesome"]}]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["css/all.css","css/v4-shims.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.14"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["bootbox"]},{"type":"character","attributes":{},"value":["4.4.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/bootbox"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["bootbox.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["localforage"]},{"type":"character","attributes":{},"value":["1.5"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/localforage"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["localforage.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial"]},{"type":"character","attributes":{},"value":["0.9.2.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial.js"]},{"type":"character","attributes":{},"value":["tutorial.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-autocompletion"]},{"type":"character","attributes":{},"value":["0.9.2.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-autocompletion.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-diagnostics"]},{"type":"character","attributes":{},"value":["0.9.2.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-diagnostics.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.9.2.1"]}]}]}
</script>
<!--/html_preserve-->
<!--html_preserve-->
<script type="application/shiny-prerendered" data-context="execution_dependencies">
{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["packages"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["packages","version"]},"class":{"type":"character","attributes":{},"value":["data.frame"]},"row.names":{"type":"integer","attributes":{},"value":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55]}},"value":[{"type":"character","attributes":{},"value":["assertthat","backports","base","colorspace","compiler","crayon","datasets","digest","dplyr","evaluate","ggplot2","glue","graphics","grDevices","grid","gridExtra","gtable","HarmonicRegression","htmltools","htmlwidgets","httpuv","jsonlite","knitr","labeling","later","lazyeval","learnr","magrittr","markdown","methods","mime","munsell","pillar","pkgconfig","promises","purrr","R6","Rcpp","rlang","rmarkdown","rprojroot","scales","shiny","stats","stringi","stringr","tibble","tidyr","tidyselect","tools","utils","withr","xfun","xtable","yaml"]},{"type":"character","attributes":{},"value":["0.2.1","1.1.4","3.6.1","1.4-1","3.6.1","1.3.4","3.6.1","0.6.20","0.8.3","0.14","3.2.0","1.3.1","3.6.1","3.6.1","3.6.1","2.3","0.3.0","1.0","0.3.6","1.3","1.5.1","1.6","1.23","0.3","0.8.0","0.2.2","0.9.2.1","1.5","1.0","3.6.1","0.7","0.5.0","1.4.2","2.0.2","1.0.1","0.3.2","2.4.0","1.0.2","0.4.0","1.14","1.3-2","1.0.0","1.3.2","3.6.1","1.4.3","1.4.0","2.1.3","0.8.3","0.2.5","3.6.1","3.6.1","2.1.2","0.8","1.8-4","2.2.0"]}]}]}
</script>
<!--/html_preserve-->
<div id="section-refs" class="references">
<div id="section-ref-hughes_harmonics_2009">
<p>Hughes, M.E., DiTacchio, L., Hayes, K.R., Vollmers, C., Pulivarthy, S., Baggs, J.E., Panda, S., and Hogenesch, J.B. (2009). Harmonics of Circadian Gene Transcription in Mammals. PLoS Genet <em>5</em>, e1000442.</p>
</div>
<div id="section-ref-hughes_jtk_cycle:_2010">
<p>Hughes, M.E., Hogenesch, J.B., and Kornacker, K. (2010). JTK_CYCLE: An Efficient Nonparametric Algorithm for Detecting Rhythmic Components in Genome-Scale Data Sets. Journal of Biological Rhythms <em>25</em>, 372–380.</p>
</div>
<div id="section-ref-hughes_guidelines_2017">
<p>Hughes, M.E., Abruzzi, K.C., Allada, R., Anafi, R., Arpat, A.B., Asher, G., Baldi, P., Bekker, C. de, Bell-Pedersen, D., Blau, J., et al. (2017). Guidelines for Genome-Scale Analysis of Biological Rhythms. Journal of Biological Rhythms <em>32</em>, 380–393.</p>
</div>
<div id="section-ref-li_considerations_2015">
<p>Li, J., Grant, G.R., Hogenesch, J.B., and Hughes, M.E. (2015). Considerations for RNA-seq Analysis of Circadian Rhythms. In Methods in Enzymology, (Elsevier), pp. 349–367.</p>
</div>
<div id="section-ref-love_rna-seq_2015">
<p>Love, M.I., Anders, S., Kim, V., and Huber, W. (2015). RNA-Seq workflow: Gene-level exploratory analysis and differential expression. F1000Research <em>4</em>, 1070.</p>
</div>
<div id="section-ref-thaben_detecting_2014">
<p>Thaben, P.F., and Westermark, P.O. (2014). Detecting Rhythms in Time Series with RAIN. Journal of Biological Rhythms 0748730414553029.</p>
</div>
<div id="section-ref-thaben_differential_2016">
<p>Thaben, P.F., and Westermark, P.O. (2016). Differential rhythmicity: Detecting altered rhythmicity in biological data. Bioinformatics <em>32</em>, 2800–2808.</p>
</div>
<div id="section-ref-wu_evaluation_2014">
<p>Wu, G., Zhu, J., Yu, J., Zhou, L., Huang, J.Z., and Zhang, Z. (2014). Evaluation of five methods for genome-wide circadian gene identification. Journal of Biological Rhythms <em>29</em>, 231–242.</p>
</div>
</div>
</div>
</div>

</div> <!-- topics -->

<div class="topicsContainer">
<div class="topicsPositioner">
<div class="band">
<div class="bandContent topicsListContainer">

<!-- begin doc-metadata -->
<div id="doc-metadata">
<h2 class="title toc-ignore" style="display:none;">Analysis of Rhythms</h2>
<h4 class="author"><em>Dr. Bharath Ananthasubramaniam | <a href="https://twitter.com/bharath_ananth">I’m on Twitter</a></em></h4>
<address class="author_afil">
Institute for Theoretical Biology, Charite Universitaetsmedizin, Berlin<br><a class="author_email" href="mailto:#"><a href="mailto:bharath.ananth@charite.de" class="email">bharath.ananth@charite.de</a></a>
</address>
<h4 class="date"><em>14 August 2019</em></h4>
</div>
<!-- end doc-metadata -->

</div> <!-- bandContent.topicsListContainer -->
</div> <!-- band -->
</div> <!-- topicsPositioner -->
</div> <!-- topicsContainer -->


</div> <!-- bandContent page -->
</div> <!-- pageContent band -->




<script>
// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});
</script>


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>

</html>
